src/target/lpc55xx.c

```diff
@@ -37,12 +37,478 @@
+		goto exit;
+	}
+
+	status = iap_call_raw(t, IAP_CMD_FFR_GET_UUID, LPC55XX_UUID_ADDRESS, 0, 0);
+
+	if (status != IAP_STATUS_FLASH_SUCCESS) {
+		DEBUG_WARN("LPC55xx: IAP error: FFR_GET_UUID (%d)\n", status);
+		goto exit;
+	}
+
+	target_mem_read(t, uuid, LPC55XX_UUID_ADDRESS, LPC55XX_UUID_LEN);
+
+	success = true;
+
+exit:
+
+	target_mem_write(t, LPC55XX_FLASH_CONFIG_ADDRESS, backup_memory, sizeof(backup_memory));
+	target_regs_write(t, regs);
+
+	return success;
+}
+
+static bool lpc55xx_flash_erase(target_flash_s *tf, target_addr_t addr, size_t len)
+{
+	enum iap_status status = iap_call_raw(tf->t, IAP_CMD_FLASH_ERASE, addr, (uint32_t)len, LPC55XX_ERASE_KEY);
+
+	if (status != IAP_STATUS_FLASH_SUCCESS) {
+		DEBUG_WARN("LPC55xx: IAP error: FLASH_ERASE (%d)\n", status);
+		return false;
+	}
+
+	return true;
+}
+
+static bool lpc55xx_flash_write(target_flash_s *tf, target_addr_t dest, const void *src, size_t len)
+{
+	target_mem_write(tf->t, LPC55XX_WRITE_BUFFER_ADDRESS, src, len);
+
+	enum iap_status	status = iap_call_raw(tf->t, IAP_CMD_FLASH_PROGRAM, dest, LPC55XX_WRITE_BUFFER_ADDRESS, (uint32_t)len);
+
+	if (status != IAP_STATUS_FLASH_SUCCESS) {
+		DEBUG_WARN("LPC55xx: IAP error: FLASH_PROGRAM (%d)\n", status);
+		return false;
+	}
+
+	return true;
+}
+
+static bool lpc55xx_enter_flash_mode(target *t)
+{
+	// TODO: do we need to reset the target here?
+	// (reset here doesn't work for some reason)
+	// target_reset(t);
+
+	lpc55xx_prepare_flash_config(t, LPC55XX_FLASH_CONFIG_ADDRESS);
+
+	enum iap_status status;
+
+	// Initialize the IAP flash context once in a predefined location
+	// of SRAM, the flash erase/write functions assume it is present.
+
+	status = iap_call_raw(t, IAP_CMD_FLASH_INIT, 0, 0, 0);
+
+	if (status != IAP_STATUS_FLASH_SUCCESS) {
+		DEBUG_WARN("LPC55xx: IAP error: FLASH_INIT (%d)\n", status);
+		return false;
+	}
+
+	return true;
+}
+
+static target_flash_s *lpc55xx_add_flash(target *t)
+{
+	target_flash_s *f = calloc(1, sizeof(*f));
+
+	if (!f) {			/* calloc failed: heap exhaustion */
+		DEBUG_WARN("calloc: failed in %s\n", __func__);
+		return NULL;
+	}
+
+	struct flash_config config;
+
+	if (!lpc55xx_flash_init(t, &config)) {
+		return NULL;
+	}
+
+	DEBUG_INFO("LPC55xx: Detected flash with %" PRIu32 " bytes, %" PRIu32 "-byte pages\n",
+		config.flash_total_size,
+		config.flash_page_size);
+
+	t->enter_flash_mode = lpc55xx_enter_flash_mode;
+	f->blocksize = config.flash_page_size;
+	f->writesize = LPC55XX_WRITE_SIZE;
+	f->start = LPC55XX_FLASH_BASE;
+	f->length = config.flash_total_size;
+	f->erase = lpc55xx_flash_erase;
+	f->write = lpc55xx_flash_write;
+	f->erased = 0xff;
+
+	target_add_flash(t, f);
+	return f;
+}
+
+static bool lpc55xx_read_uid(target *t, int argc, const char *argv[])
+{
+	(void)argc;
+	(void)argv;
+
+	uint8_t uuid[LPC55XX_UUID_LEN];
+
+	if (!lpc55xx_get_uuid(t, uuid))
+		return false;
+
+	tc_printf(t, "UID: 0x");
+	for (uint32_t i = 0; i < sizeof(uuid); ++i)
```

Comment:
Please use size_t here - sizeof() returns a size_t value and you're using i for array indexing.

Reply:
Agreed, I actually just copied this from the lpc11xx readuid